_posts/2025-04-11-Hybrid-Search-Options-And-Practice.md

```diff
@@ -0,0 +1,275 @@
+
+API Request:
+
+
+```
+GET /bookstore_catalog/_search
+{
+  "query": {
+    "match": {
+      "title": "the martian"
+    }
+  }
+}
+```
+
+
+
+#### Semantic Search
+
+
+Semantic search in OpenSearch represents a significant advancement in search technology, leveraging the power of machine learning and natural language processing. This approach uses dense vector embeddings to represent both documents and queries in a high-dimensional space, allowing for semantic similarity comparisons that go beyond simple keyword matching. At the core of semantic search is the concept of text embeddings - dense vector representations of text that capture semantic meaning. These embeddings are typically generated using large language models trained on vast corpora of text. In OpenSearch, these embeddings are stored as dense vector fields and can be efficiently searched using k nearest neighbor algorithms. At search time, the user's query is similarly transformed into a vector embedding. The system then finds documents whose vector representations are closest to the query vector, typically using cosine similarity or Euclidean distance.
+
+API Request:
+
+
+```
+GET /bookstore_catalog/_search
+{
+  "_source": {
+    "excludes": [
+      "passage_embedding"
+    ]
+  },
+  "query": {
+    "neural": {
+      "title_embedding": {
+        "query_text": "science fiction",
+        "k": 100
+      }
+    }
+  }
+}
+```
+
+This query uses:
+
+* `title_embedding`: The query vector of the field title.
+* `k`: Number of nearest neighbors to retrieve
+
+In summary, lexical search algorithms like BM25 provide precise keyword matching. They are fast and efficient but do not incorporate semantic meaning. On the other hand, semantic search understands meaning and context, handling natural language queries well. However, it may miss important keywords, especially for fact-based queries. As we enter the era of Generative AI, an increasing number of use cases require both keyword matching and semantic understanding. Hybrid search is OpenSearch's solution to address this need, combining the strengths of both approaches.
+
+### 2. Hybrid Search: Solving the Problem and Options
+
+Hybrid search often involves combining scores from different search methods (e.g., BM25 for keyword search and semantics for vector search). However, these scores can have vastly different scales and distributions. Normalization helps to bring these scores into a common range, allowing for fair comparison and combination. OpenSearch offers various normalization techniques that can be broadly categorized into two main types: Score-based and Rank-based. Let's explore these categories and their specific techniques in detail.
```

Comment:
can you add example of such queries where scores are at different scales, e.g. knn with L2 space vs text match query, or even within one query like knn with L2 and inner product when vectors are not normalized (latest can be unbounded, actual scores can be [-200, .. +200]